File: src/utils/coursePolicy.js
import { formatEventActor } from './calendarEvents';

export { formatEventActor };

export const POLICY_FIELDS = [
    { key: 'markingScheme', label: 'Marking scheme', placeholder: 'Midsem, endsem, quiz weights…' },
    { key: 'attendancePolicy', label: 'Attendance policy', placeholder: 'Minimum attendance, penalty rules…' },
    { key: 'auditWithdrawalPolicy', label: 'Audit / withdrawal policy', placeholder: 'Audit criteria, drop deadlines…' },
    { key: 'otherNotes', label: 'Other notes', placeholder: 'Textbooks, TA contacts, links…' },
];

export function emptyPolicyDraft(policy = null) {
    return {
        markingScheme: policy?.markingScheme || '',
        attendancePolicy: policy?.attendancePolicy || '',
        auditWithdrawalPolicy: policy?.auditWithdrawalPolicy || '',
        otherNotes: policy?.otherNotes || '',
        createdBy: policy?.createdBy || null,
        updatedBy: policy?.updatedBy || null,
    };
}

function trimField(value) {
    if (value == null) return '';
    return String(value).trim();
}

export function isPolicySubmittable(draft) {
    if (!draft || typeof draft !== 'object') return false;
    return POLICY_FIELDS.some(({ key }) => trimField(draft[key]).length > 0);
}

export function policyPayload(draft) {
    const out = {};
    for (const { key } of POLICY_FIELDS) {
        out[key] = trimField(draft[key]);
    }
    return out;
}

export function actorsMatch(a, b) {
    if (!a || !b) return false;
    return (
        a.at === b.at &&
        (a.kerberos || a.name) === (b.kerberos || b.name)
    );
}

export function policyHasContent(policy) {
    if (!policy) return false;
    return POLICY_FIELDS.some(({ key }) => trimField(policy[key]).length > 0);
}

export function formatPolicyActorLine(actor) {
    const info = formatEventActor(actor);
    if (!info) return null;
    return `${info.who} · ${info.when}`;
}

File: app/lib/core/course_policy.dart
// Validation and payload helpers for course policy forms.
// Mirrors src/utils/coursePolicy.js.

class PolicyFieldDef {
  const PolicyFieldDef(this.key, this.label, this.placeholder);

  final String key;
  final String label;
  final String placeholder;
}

const kPolicyFields = [
  PolicyFieldDef('markingScheme', 'Marking scheme', 'Midsem, endsem, quiz weights…'),
  PolicyFieldDef('attendancePolicy', 'Attendance policy', 'Minimum attendance, penalty rules…'),
  PolicyFieldDef(
    'auditWithdrawalPolicy',
    'Audit / withdrawal policy',
    'Audit criteria, drop deadlines…',
  ),
  PolicyFieldDef('otherNotes', 'Other notes', 'Textbooks, TA contacts, links…'),
];

String _trim(String? value) => (value ?? '').trim();

bool isPolicySubmittable(CoursePolicyDraftLike draft) {
  return kPolicyFields.any((f) => _trim(draftField(draft, f.key)).isNotEmpty);
}

String draftField(CoursePolicyDraftLike draft, String key) {
  switch (key) {
    case 'markingScheme':
      return draft.markingScheme;
    case 'attendancePolicy':
      return draft.attendancePolicy;
    case 'auditWithdrawalPolicy':
      return draft.auditWithdrawalPolicy;
    case 'otherNotes':
      return draft.otherNotes;
    default:
      return '';
  }
}

void setDraftField(CoursePolicyDraftLike draft, String key, String value) {
  if (key == 'markingScheme') {
    draft.markingScheme = value;
  } else if (key == 'attendancePolicy') {
    draft.attendancePolicy = value;
  } else if (key == 'auditWithdrawalPolicy') {
    draft.auditWithdrawalPolicy = value;
  } else if (key == 'otherNotes') {
    draft.otherNotes = value;
  }
}

Map<String, dynamic> policyPayload(CoursePolicyDraftLike draft) => {
      'markingScheme': _trim(draft.markingScheme),
      'attendancePolicy': _trim(draft.attendancePolicy),
      'auditWithdrawalPolicy': _trim(draft.auditWithdrawalPolicy),
      'otherNotes': _trim(draft.otherNotes),
    };

abstract class CoursePolicyDraftLike {
  String get markingScheme;
  set markingScheme(String value);
  String get attendancePolicy;
  set attendancePolicy(String value);
  String get auditWithdrawalPolicy;
  set auditWithdrawalPolicy(String value);
  String get otherNotes;
  set otherNotes(String value);
}
